docs/pages/machine-id/guides/gitlab.mdx

```diff
@@ -0,0 +1,164 @@
+- `namespace_path`: the group or user that the project belongs to.
+- `project_path`: the full name of the project that the run belongs to. For
+example: `mygroup/myproject`
+- `pipeline_source`: the trigger of the CI run. For example: `web`
+- `environment`: the environment that the job is associated with
+
+A full explanation of these constraints values can be found on the [GitLab
+Documentation.](https://docs.gitlab.com/ee/ci/secrets/id_token_authentication.html#token-payload)
+
+In this example, you will create a rule that grants access to any GitLab CI job
+within a specific GitLab project. Determine the fully qualified path of your
+GitLab project. This will include your username (or group) and the name of your
+project, e.g `my-user/my-project`.
+
+Create a file named `gitlab-token.yaml`. Ensure you substitute any values as
+suggested by the comments in this example.
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: gitlab-demo
+spec:
+  roles: [Bot]
+  join_method: gitlab
+  bot_name: gitlab-demo
+  gitlab:
+    # domain should be the domain of your GitLab instance. If you are using
+    # GitLab's cloud hosted offering, omit this field entirely.
+    domain: gitlab.example.com
+    allow:
+        # project_path should be the fully qualified path of your GitLab
+        # project that you determined earlier.
+      - project_path: my-user/my-project
+```
+
+Apply this to your Teleport cluster using `tctl`:
+
+```code
+$ tctl create -f gitlab-token.yaml
+```
+
+## Step 2/3. Create a Machine ID bot
+
+With the join token created, the next step is to create the Machine ID bot that
+the token will grant access to.
+
+In this example, the bot is given the default `access` role. In a production
+environment, the principle of least privilege should be applied and you should
+create a role that grants the bot access to the precise resources that will be
+needed in your CI/CD pipeline.
+
+Enter the following command from your workstation with `tsh` access, replacing
+the `ci-user` login with a Linux user on the host that you want your GitLab
+pipeline to be able to connect to:
+
+```code
+$ tctl bots add gitlab-demo --roles=access --token=gitlab-demo --logins=ci-user
+```
+
+## Step 3/3. Configure a GitLab Pipeline
+
+With the bot and join token created, you can now create a GitLab pipeline that
+uses these to access a node in your cluster.
+
+Create a file called `.gitlab-ci.yml` in the root of your repository. Replace:
+- `teleport.example.com` with the name of your Teleport cluster
+- `teleport.example.com:443` with the public address of your Teleport Proxy
+- `my-node` and `ci-user` with the user and host you wish to connect to
+
+```yaml
+stages:
+  - deploy
+
+deploy-job:
+  stage: deploy
+  id_tokens:
+    TBOT_GITLAB_JWT:
+      # An environment variable named TBOT_GITLAB_JWT must exist and contain
+      # an ID token with an audience that matches your Teleport cluster's name.
+      aud: teleport.example.com
```

Comment:
This was a bit confusing and difficult to track down in the docs. Including a link around here would be helpful (e.g. https://docs.gitlab.com/ee/ci/secrets/id_token_authentication.html).

It also might be worth adding a note somewhere about how this token is used. I think it may be being used by the Teleport binary itself to confirm that it's being run from GitLab CI?

Reply:
It doesn't confirm it's being run from GitLab CI, but confirms that the generated ID token is intended for our Teleport cluster. Without an audience, if the user sent an ID token to GCP in order to try and complete some action on GCP, and if GCP was compromised, this ID token could then be used against the user's Teleport cluster.

I'll add some better words explaining that/linking to the docs.

Reply:
Yeah, that description is largely correct bar one point

> We give the --token=gitlab-demo to this user, which registers this user as being connected to the properties specified in the token generated in step 1.

This isn't exactly true. This is actually just a form of additional validation. The binding of the token to a bot exists only within the token, but we require that a token name is provided when creating a bot so we can provide a command line feedback to the user if the token is configured incorrectly. It's a bit of an awkward order, but we could change this in future.

Reply:
c15e7d7cc4 link added